packages/next-routing/src/types.ts

```diff
@@ -0,0 +1,83 @@
+export type RouteHas =
+  | {
+      type: 'header' | 'cookie' | 'query'
+      key: string
+      value?: string
+    }
+  | {
+      type: 'host'
+      key?: undefined
+      value: string
+    }
+
+export type Route = {
+  // regex as string can have named or un-named matches
+  sourceRegex: string
+  // destination can have matches to replace in destination
+  // keyed by $1 for un-named and $name for named
+  destination: string
+  headers?: Record<string, string>
+  has?: RouteHas[]
+  missing?: RouteHas[]
+  status?: number
+}
+
+export type MiddlewareContext = {
+  url: URL
+  headers: Headers
+  requestBody: ReadableStream
+}
+
+export type MiddlewareResult = {
+  bodySent?: boolean
+  requestHeaders?: Headers
+  redirect?: {
+    url: URL
+    status: number
+  }
+  rewrite?: URL
+}
+
+export type ResolveRoutesParams = {
+  url: URL
+  basePath: string
+  requestBody: ReadableStream
+  headers: Headers
+  pathnames: string[]
+  i18n?: {
+    defaultLocale: string
+    domains?: Array<{
+      defaultLocale: string
+      domain: string
+      http?: true
+      locales?: string[]
+    }>
+    localeDetection?: false
+    locales: string[]
+  }
+  routes: {
+    beforeMiddleware: Array<Route>
+    beforeFiles: Array<Route>
+    afterFiles: Array<Route>
+    dynamicRoutes: Array<Route>
+    onMatch: Array<Route>
+    fallback: Array<Route>
+  }
```

Comment:
Is it intentional for `routes` type to not have same shape as `routes` provided in `onBuildComplete` adapter callback? 

If so - how should one convert those to match expected shape here? Concat adapter's `headers` and `redirects` and put those in `beforeMiddleware`? Or should it be redirects there and `headers` in `onMatch`?